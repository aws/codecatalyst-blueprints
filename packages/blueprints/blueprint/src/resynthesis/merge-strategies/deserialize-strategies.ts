import * as fs from 'fs';
import * as path from 'path';
import { Strategy } from './models';
import { Ownership } from '../ownership';
import { walkFiles } from '../walk-files';

export type StrategyLocations = { [bundlePath: string]: Strategy[] };

const getStrategyIds = (strategies: StrategyLocations): { [identifier: string]: Strategy } => {
  const ids: { [id: string]: Strategy } = {};
  for (const strategyList of Object.values(strategies)) {
    for (const strategy of strategyList) {
      ids[strategy.identifier] = strategy;
    }
  }
  return ids;
};

export const deserializeStrategies = (existingBundle: string, strategyMatch: StrategyLocations): StrategyLocations => {
  const inMemStrategies = getStrategyIds(strategyMatch);
  const validStrategies: StrategyLocations = {};
  const ownershipFiles = walkFiles(existingBundle, [`src/**/*${Ownership.DEFAULT_FILE_NAME}`]);

  ownershipFiles.forEach(ownerfile => {
    const ownershipObject = Ownership.asObject(fs.readFileSync(path.join(existingBundle, ownerfile)).toString(), inMemStrategies);
    const relevantStrategies: Strategy[] = [];
    ownershipObject.resynthesis?.strategies.forEach(deserializedStrategy => {
      if (deserializedStrategy.identifier in inMemStrategies) {
        relevantStrategies.push(deserializedStrategy);
      }
    });
    validStrategies[ownerfile] = relevantStrategies;
  });

  return validStrategies;
};
