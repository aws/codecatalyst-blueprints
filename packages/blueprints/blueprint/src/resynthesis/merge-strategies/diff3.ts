import { diff_match_patch, Diff, DIFF_DELETE, DIFF_EQUAL, DIFF_INSERT } from 'diff-match-patch';

const DMP = new diff_match_patch();
const CONFLICT_MARKER_LENGTH = 7;

function diffLines(a: string, b: string, dmp: diff_match_patch = DMP): Diff[] {
  const lines = dmp.diff_linesToChars_(a, b);
  const diff = dmp.diff_main(lines.chars1, lines.chars2, false);
  dmp.diff_charsToLines_(diff, lines.lineArray);

  dmp.diff_cleanupSemantic(diff);
  return diff;
}

type MergeChunk = MergeConflictChunk | MergeOkChunk;

interface MergeConflictChunk {
  conflict: true;
  a: string;
  b: string;
}

interface MergeOkChunk {
  conflict: false;
  ok: string;
}

export interface Diff3Options {
  diffFunction?: (a: string, b: string) => Diff[];
  aLabel?: string;
  bLabel?: string;
}

export class Diff3 {
  private readonly lines: {
    a: string[];
    o: string[];
    b: string[];
  };

  private currentLine: {
    a: number;
    o: number;
    b: number;
  } = {
      a: 0,
      o: 0,
      b: 0,
    };

  private readonly matches: {
    a: Map<number, number>;
    b: Map<number, number>;
  };

  private readonly mergedChunks: MergeChunk[] = [];
  private readonly options?: Diff3Options;

  constructor(a: string, o: string, b: string, options?: Diff3Options) {
    this.options = options;
    const diff = options?.diffFunction ?? diffLines;

    this.matches = {
      a: getMatches(diff(o, a)),
      b: getMatches(diff(o, b)),
    };

    this.lines = {
      a: splitLines(a),
      o: splitLines(o),
      b: splitLines(b),
    };

    this.merge();
  }

  getMerged(): string {
    return this.mergedChunks
      .map(chunk =>
        mergeChunkToString(chunk, {
          aLabel: this.options?.aLabel,
          bLabel: this.options?.bLabel,
        }),
      )
      .join('');
  }

  // Based on the algorithm described in https://www.cis.upenn.edu/~bcpierce/papers/diff3-short.pdf
  private merge() {
    while (true) {
      const i = this.findMismatch();

      if (i === 1) {
        const { o, a, b } = this.findMatch();

        if (a !== undefined && b !== undefined) {
          this.pushChunk(a, o, b);
        } else {
          this.pushRemainingChunk();
          return;
        }
      } else if (i) {
        this.pushChunk(this.currentLine.a + i, this.currentLine.o + i, this.currentLine.b + i);
      } else {
        this.pushRemainingChunk();
        return;
      }
    }
  }

  /**
   * Find the offset from the current line to the next mismatch
   * @returns the number of lines to the next mismatch, or undefined if there is no mismatch
   */
  private findMismatch(): number | undefined {
    let i = 1;

    while (
      this.currentLine.a + i <= this.lines.a.length ||
      this.currentLine.b + i <= this.lines.b.length ||
      this.currentLine.o + i <= this.lines.o.length
    ) {
      if (
        this.matches.a.get(this.currentLine.o + i) !== this.currentLine.a + i ||
        this.matches.b.get(this.currentLine.o + i) !== this.currentLine.b + i
      ) {
        return i;
      }

      i++;
    }

    return;
  }

  private findMatch(): {
    a: number | undefined;
    o: number;
    b: number | undefined;
  } {
    let o = this.currentLine.o + 1;

    while (o <= this.lines.o.length && this.matches.a.get(o) === undefined && this.matches.b.get(o) === undefined) {
      o++;
    }

    return { o, a: this.matches.a.get(o), b: this.matches.b.get(o) };
  }

  private pushRemainingChunk() {
    this.pushChunk(this.lines.a.length + 1, this.lines.o.length + 1, this.lines.b.length + 1);
  }

  private pushChunk(a: number, o: number, b: number) {
    const oText = this.lines.o.slice(this.currentLine.o, o - 1).join('');
    const aText = this.lines.a.slice(this.currentLine.a, a - 1).join('');
    const bText = this.lines.b.slice(this.currentLine.b, b - 1).join('');

    if (oText === aText && oText === bText) {
      this.mergedChunks.push({
        conflict: false,
        ok: oText,
      });
    } else if (oText === aText) {
      this.mergedChunks.push({
        conflict: false,
        ok: bText,
      });
    } else if (oText === bText) {
      this.mergedChunks.push({
        conflict: false,
        ok: aText,
      });
    } else if (aText === bText) {
      // false conflict, both copies modified the original in the same way.
      this.mergedChunks.push({
        conflict: false,
        ok: aText,
      });
    } else {
      this.mergedChunks.push({
        conflict: true,
        a: aText,
        b: bText,
      });
    }

    this.currentLine = {
      o: o - 1,
      a: a - 1,
      b: b - 1,
    };
  }
}

/**
 * Converts a diff to a map of matching lines between the original and modified text.
 * @param diff A diff-match-patch style diff
 * @returns a map from line number in the original text to line number in the modified text.
 */
function getMatches(diff: Diff[]): Map<number, number> {
  const matches = new Map<number, number>();

  let oLineNumber = 1;
  let aLineNumber = 1;
  for (const [type, text] of diff) {
    const numLines = countLines(text);

    switch (type) {
      case DIFF_EQUAL:
        for (let i = 0; i < numLines; i++) {
          matches.set(oLineNumber + i, aLineNumber + i);
        }

        oLineNumber += numLines;
        aLineNumber += numLines;
        break;
      case DIFF_DELETE:
        oLineNumber += numLines;
        break;
      case DIFF_INSERT:
        aLineNumber += numLines;
        break;
    }
  }

  return matches;
}

function countLines(s: string): number {
  let count = 0;
  for (let i = 0; i < s.length; i++) {
    if (s.charAt(i) === '\n') {
      count++;
    }
  }

  return count;
}

function mergeChunkToString(
  chunk: MergeChunk,
  options?: {
    aLabel?: string;
    bLabel?: string;
  },
) {
  if (chunk.conflict) {
    let chunkContents = '';

    chunkContents += createConflictMarker('<', options?.aLabel);
    chunkContents += chunk.a;
    if (!chunk.a.endsWith('\n')) {
      chunkContents += '\n';
    }

    chunkContents += createConflictMarker('=');
    chunkContents += chunk.b;
    if (!chunk.b.endsWith('\n')) {
      chunkContents += '\n';
    }
    chunkContents += createConflictMarker('>', options?.bLabel);

    return chunkContents;
  } else {
    return chunk.ok;
  }
}

function createConflictMarker(type: '<' | '=' | '>', label?: string): string {
  return type.repeat(CONFLICT_MARKER_LENGTH) + `${label ? ' ' + label : ''}\n`;
}

/**
 * Splits a string into lines, preserving the newline character at the end of each line.
 * @param s the string to split
 * @returns an array of lines
 */
function splitLines(s: string): string[] {
  const split = s.split('\n');

  return split.map((line, i) => (i === split.length - 1 ? line : line + '\n'));
}
