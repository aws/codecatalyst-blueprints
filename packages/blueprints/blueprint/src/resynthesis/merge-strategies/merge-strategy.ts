import { ContextFile } from './models';

type StrategyFunction = (
  _commonAncestorFile: ContextFile | undefined,
  _existingFile: ContextFile | undefined,
  _proposedFile: ContextFile | undefined,
  _options?: {}) => ContextFile | undefined;

export class MergeStrategies {
  public static alwaysUpdate: StrategyFunction = function alwaysUpdate(
    _commonAncestorFile: ContextFile | undefined,
    _existingFile: ContextFile | undefined,
    _proposedFile: ContextFile | undefined,
    _options?: {}) {
    return _existingFile;
  };

  public static neverUpdate: StrategyFunction = function neverUpdate(
    _commonAncestorFile: ContextFile | undefined,
    _existingFile: ContextFile | undefined,
    _proposedFile: ContextFile | undefined,
    _options?: {}) {
    return _existingFile;
  };
}
