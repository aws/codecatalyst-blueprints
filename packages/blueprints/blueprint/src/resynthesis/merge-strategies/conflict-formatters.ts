import { CONFLICT_MARKER_LENGTH, splitLines } from './diff3';

export type MergeConflictFormatter = (a: string, o: string, b: string, options?: MergeConflictFormatterOptions) => string;

export interface MergeConflictFormatterOptions {
  aLabel?: string;
  oLabel?: string;
  bLabel?: string;
}

export class ConflictFormatters {
  public static diff3(a: string, o: string, b: string, options?: MergeConflictFormatterOptions): string {
    return createConflictString(a, b, {
      oText: o,
      aLabel: options?.aLabel,
      oLabel: options?.oLabel,
      bLabel: options?.bLabel,
    });
  }

  public static diff3NoAncestor(a: string, _o: string, b: string, options?: MergeConflictFormatterOptions) {
    return createConflictString(a, b, {
      aLabel: options?.aLabel,
      bLabel: options?.bLabel,
    });
  }

  public static preferExisting(a: string, _o: string, _b: string): string {
    return a;
  }

  public static preferProposed(_a: string, _o: string, b: string): string {
    return b;
  }

  public static trimEnds(a: string, _o: string, b: string, options?: MergeConflictFormatterOptions): string {
    const aLines = splitLines(a);
    const bLines = splitLines(b);

    const resultLines: string[] = [];
    const minLength = Math.min(aLines.length, bLines.length);
    for (let i = 0; i < minLength; i++) {
      if (aLines[i] !== bLines[i]) {
        break;
      }

      resultLines.push(aLines[i]);
    }

    const commonPrefixLength = resultLines.length;
    let commonSuffix: string[] = [];
    for (let i = 0; i < minLength - commonPrefixLength; i++) {
      if (aLines[aLines.length - 1 - i] !== bLines[bLines.length - 1 - i]) {
        break;
      }

      commonSuffix.push(aLines[aLines.length - 1 - i]);
    }

    const aConflictLines = aLines.slice(commonPrefixLength, aLines.length - commonSuffix.length);
    const bConflictLines = bLines.slice(commonPrefixLength, bLines.length - commonSuffix.length);
    if (aConflictLines.length || bConflictLines.length) {
      resultLines.push(createConflictString(aConflictLines.join(''), bConflictLines.join(''), { ...options }));
    }

    if (commonSuffix.length) {
      resultLines.push(...commonSuffix.reverse());
    }

    return resultLines.join('');
  }
}

function createConflictString(
  aText: string,
  bText: string,
  options?: {
    aLabel?: string;
    bLabel?: string;
    oText?: string;
    oLabel?: string;
  },
) {
  let chunkContents = '';

  chunkContents += createConflictMarker('<', options?.aLabel);
  chunkContents += aText;
  if (!aText.endsWith('\n')) {
    chunkContents += '\n';
  }

  if (options?.oText !== undefined) {
    chunkContents += createConflictMarker('|', options?.oLabel);
    chunkContents += options.oText;
    if (!options.oText.endsWith('\n')) {
      chunkContents += '\n';
    }
  }

  chunkContents += createConflictMarker('=');
  chunkContents += bText;
  if (!bText.endsWith('\n')) {
    chunkContents += '\n';
  }
  chunkContents += createConflictMarker('>', options?.bLabel);

  return chunkContents;
}

function createConflictMarker(type: '<' | '|' | '=' | '>', label?: string): string {
  return type.repeat(CONFLICT_MARKER_LENGTH) + `${label ? ' ' + label : ''}\n`;
}
