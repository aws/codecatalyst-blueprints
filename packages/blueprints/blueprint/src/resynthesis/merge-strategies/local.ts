import { execSync } from 'child_process';
import { randomBytes } from 'crypto';
import { mkdirSync, readFileSync, rmSync, writeFileSync } from 'fs';
import * as path from 'path';
import { CONFLICT_MARKER_LENGTH } from './diff3';
import { StrategyFunction } from './merge-strategies';
import { ContextFile } from '../context-file';

export function constructLocalStrategy(owner: string, ownershipPath: string): StrategyFunction {
  return (
    commonAncestorFile: ContextFile | undefined,
    existingFile: ContextFile | undefined,
    proposedFile: ContextFile | undefined,
  ): ContextFile | undefined => {
    const repositoryName = existingFile?.repositoryName ?? proposedFile?.repositoryName ?? commonAncestorFile?.repositoryName;
    if (!repositoryName) {
      throw new Error('Failed to determine repository name because no input files were provided.');
    }

    const resolvedPath = existingFile?.path ?? proposedFile?.path ?? commonAncestorFile?.path;
    if (!resolvedPath) {
      throw new Error('Failed to determine path because no input files were provided.');
    }

    const resolvedBuffer = runLocalCommand(owner, ownershipPath, resolvedPath, {
      a: existingFile?.buffer ?? Buffer.from(''),
      o: commonAncestorFile?.buffer ?? Buffer.from(''),
      b: proposedFile?.buffer ?? Buffer.from(''),
    });
    if (!resolvedBuffer) {
      return undefined;
    }

    return {
      repositoryName,
      path: resolvedPath,
      buffer: resolvedBuffer,
    };
  };
}

function runLocalCommand(
  owner: string,
  ownershipPath: string,
  resolvedPath: string,
  buffers: { a: Buffer; o: Buffer; b: Buffer },
): Buffer | undefined {
  createTempDir();

  let cmd: string | undefined;
  try {
    const commonTempFile = createTempFile(buffers.o);
    const existingTempFile = createTempFile(buffers.a);
    const proposedTempFile = createTempFile(buffers.b);

    const cwd = path.dirname(ownershipPath);
    cmd = formatLocalCommand(
      owner,
      path.relative(cwd, existingTempFile),
      path.relative(cwd, commonTempFile),
      path.relative(cwd, proposedTempFile),
      resolvedPath,
    );

    execSync(cmd, {
      cwd,
    });

    // git merge driver conventions expect the resolved file to have been written to the existing file's path:
    return getResolvedFile(existingTempFile);
  } catch (error: unknown) {
    let message = 'Failed to run local merge strategy';
    if (cmd) {
      message += `: ${cmd}`;
    }

    throw new Error(`${message}: ${error}`);
  } finally {
    deleteTempDir();
  }
}

function getResolvedFile(resolvedPath: string): Buffer | undefined {
  try {
    return readFileSync(resolvedPath);
  } catch (error: any) {
    if (error.code === 'ENOENT') {
      return undefined;
    }

    throw error;
  }
}

function createTempDir() {
  try {
    mkdirSync(path.join('synth', 'tmp'), { recursive: true });
  } catch (error: any) {
    if (error.code === 'EEXIST') {
      return;
    }

    throw error;
  }
}

function deleteTempDir() {
  rmSync(path.join('synth', 'tmp'), { recursive: true });
}

function createTempFile(contents: Buffer): string {
  const randomSuffix = randomBytes(8).toString('hex');
  const tempPath = path.join('synth', 'tmp', `.merge_file_${randomSuffix}`);
  writeFileSync(tempPath, contents, {
    flag: 'wx',
  });

  return tempPath;
}

function formatLocalCommand(commandPattern: string, aPath: string, oPath: string, bPath: string, resolvedPath: string): string {
  // see: https://git-scm.com/docs/gitattributes#_defining_a_custom_merge_driver
  return commandPattern
    .replace(/%O/g, oPath)
    .replace(/%A/g, aPath)
    .replace(/%B/g, bPath)
    .replace(/%P/g, resolvedPath)
    .replace(/%L/g, `${CONFLICT_MARKER_LENGTH}`);
}
