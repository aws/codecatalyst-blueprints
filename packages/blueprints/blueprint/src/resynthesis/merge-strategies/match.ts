import * as path from 'path';
import { MergeStrategies } from './merge-strategies';
import { Strategy } from './models';
import { Ownership } from '../ownership';
import { matchesGlob } from '../walk-files';

export const FALLBACK_STRATEGY = MergeStrategies.preferProposed;
export const FALLBACK_STRATEGY_ID = `FALLBACK_${FALLBACK_STRATEGY.name}`;

export function match(sourceCodePath: string, strategies: { [bundlepath: string]: Strategy[] }): Strategy {
  const directories = sourceCodePath.split('/');

  while (directories.length > 0) {
    directories.pop();

    // path we would expect for if the ownership is declared in memory at this location
    const syntheticPath = path.join(...directories);
    // path we would expect for a ownership override file at this location
    const ownershipPath = path.join(syntheticPath, Ownership.DEFAULT_FILE_NAME);

    const commonPath = directories.join('/') + '/';
    const relativeStrategies = [...(strategies[syntheticPath] || []), ...(strategies[ownershipPath] || [])];
    if (!relativeStrategies.length) {
      continue;
    }

    const relativeBundlePath = sourceCodePath.startsWith(commonPath) ? sourceCodePath.slice(commonPath.length) : sourceCodePath;
    const matchedStrategy = matchStrategies(relativeBundlePath, relativeStrategies);
    if (matchedStrategy) {
      return matchedStrategy;
    }
  }

  return {
    identifier: FALLBACK_STRATEGY_ID,
    strategy: FALLBACK_STRATEGY,
    globs: ['*'],
  };
}

function matchStrategies(bundlePath: string, strategies: Strategy[]): Strategy | undefined {
  for (let i = strategies.length - 1; i >= 0; i--) {
    const strategy = strategies[i];
    if (matchesGlob(bundlePath, strategy.globs)) {
      return strategy;
    }
  }
  return;
}
