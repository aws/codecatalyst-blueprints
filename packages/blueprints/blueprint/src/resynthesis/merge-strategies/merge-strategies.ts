import { ConflictFormatters, MergeConflictFormatter } from './conflict-formatters';
import { Diff3 } from './diff3';
import { ContextFile } from '../context-file';

export type StrategyFunction = (
  commonAncestorFile: ContextFile | undefined,
  existingFile: ContextFile | undefined,
  proposedFile: ContextFile | undefined,
  options?: {},
) => ContextFile | undefined;

export class MergeStrategies {
  /**
   * A strategy that always resolves to the proposed file.
   * @returns the proposed file.
   */
  public static alwaysUpdate: StrategyFunction = function alwaysUpdate(
    _commonAncestorFile: ContextFile | undefined,
    _existingFile: ContextFile | undefined,
    proposedFile: ContextFile | undefined,
    _options?: {},
  ) {
    return proposedFile;
  };

  /**
   * A strategy that always resolves to the existing file.
   * @returns the existing file.
   */
  public static neverUpdate: StrategyFunction = function neverUpdate(
    _commonAncestorFile: ContextFile | undefined,
    existingFile: ContextFile | undefined,
    _proposedFile: ContextFile | undefined,
    _options?: {},
  ) {
    return existingFile;
  };

  /**
   * A strategy that resolves to the proposed file when an existing file does not exist already.
   * Otherwise, resolves to the existing file.
   */
  public static onlyAdd: StrategyFunction = function onlyAdd(
    _commonAncestorFile: ContextFile | undefined,
    existingFile: ContextFile | undefined,
    proposedFile: ContextFile | undefined,
    _options?: {},
  ) {
    return existingFile ? existingFile : proposedFile;
  };

  /**
   * A strategy that performs a three way merge between the existing, proposed, and common
   * ancestor files. The resolved file may contain conflict markers if the files can not be
   * cleanly merged.
   *
   * The provided files' contents must be UTF-8 encoded in order for the strategy to produce
   * meaningful output. The strategy attempts to detect if the input files are binary. If
   * the strategy detects a merge conflict in a binary file, it always returns the
   * proposedFile.
   */
  public static threeWayMerge: StrategyFunction = function threeWayMerge(
    commonAncestorFile: ContextFile | undefined,
    existingFile: ContextFile | undefined,
    proposedFile: ContextFile | undefined,
  ) {
    return merge(commonAncestorFile, existingFile, proposedFile);
  };

  /**
   * A strategy that performs a three way merge between the existing, proposed, and common
   * ancestor files. This strategy resolves conflicts by selecting the proposedFile's side
   * of each conflict.
   */
  public static preferProposed: StrategyFunction = function preferProposed(
    commonAncestorFile: ContextFile | undefined,
    existingFile: ContextFile | undefined,
    proposedFile: ContextFile | undefined,
  ) {
    return merge(commonAncestorFile, existingFile, proposedFile, {
      conflictFormatter: ConflictFormatters.preferProposed,
      binaryConflictPreference: 'proposed',
    });
  };

  /**
   * A strategy that performs a three way merge between the existing, proposed, and common
   * ancestor files. This strategy resolves conflicts by selecting the existingFile's side
   * of each conflict.
   */
  public static preferExisting: StrategyFunction = function preferExisting(
    commonAncestorFile: ContextFile | undefined,
    existingFile: ContextFile | undefined,
    proposedFile: ContextFile | undefined,
  ) {
    return merge(commonAncestorFile, existingFile, proposedFile, {
      conflictFormatter: ConflictFormatters.preferExisting,
      binaryConflictPreference: 'existing',
    });
  };
}

function merge(
  commonAncestorFile: ContextFile | undefined,
  existingFile: ContextFile | undefined,
  proposedFile: ContextFile | undefined,
  options?: {
    binaryConflictPreference?: 'existing' | 'proposed';
    conflictFormatter?: MergeConflictFormatter;
  },
) {
  const conflictFormatter = options?.conflictFormatter ?? ConflictFormatters.trimEnds;
  const binaryConflictPreference = options?.binaryConflictPreference ?? 'proposed';

  // The three way merge algorithm below is text based and won't produce a meaningful result
  // when merging binary files. Handle binary file merges separately:
  if (isBinary(proposedFile?.buffer) || isBinary(existingFile?.buffer) || isBinary(commonAncestorFile?.buffer)) {
    return mergeBinaryFile(commonAncestorFile, existingFile, proposedFile, { binaryConflictPreference });
  }

  if (!existingFile && proposedFile && commonAncestorFile && proposedFile.buffer.equals(commonAncestorFile.buffer)) {
    return undefined;
  }
  if (!proposedFile && existingFile && commonAncestorFile && existingFile.buffer.equals(commonAncestorFile.buffer)) {
    return undefined;
  }
  if (!existingFile && !proposedFile) {
    return undefined;
  }

  const diff3 = new Diff3(existingFile?.buffer.toString() ?? '', commonAncestorFile?.buffer.toString() ?? '', proposedFile?.buffer.toString() ?? '', {
    aLabel: 'existing',
    bLabel: 'proposed',
    conflictFormatter,
  });

  const repositoryName = existingFile?.repositoryName ?? proposedFile?.repositoryName ?? commonAncestorFile?.repositoryName;
  if (!repositoryName) {
    throw new Error('Failed to determine repository name because no input files were provided.');
  }

  const path = existingFile?.path ?? proposedFile?.path ?? commonAncestorFile?.path;
  if (!path) {
    throw new Error('Failed to determine path because no input files were provided.');
  }

  return {
    repositoryName,
    path,
    buffer: Buffer.from(diff3.getMerged()),
  };
}

function mergeBinaryFile(
  commonAncestorFile: ContextFile | undefined,
  existingFile: ContextFile | undefined,
  proposedFile: ContextFile | undefined,
  options: {
    binaryConflictPreference: 'existing' | 'proposed';
  },
): ContextFile | undefined {
  const existingIsUnchanged = buffersEqual(commonAncestorFile?.buffer, existingFile?.buffer);
  const proposedIsUnchanged = buffersEqual(commonAncestorFile?.buffer, proposedFile?.buffer);

  if (existingIsUnchanged && proposedIsUnchanged) {
    // no conflict - all three equal
    return commonAncestorFile;
  } else if (existingIsUnchanged) {
    // no conflict - take changes from proposed
    return proposedFile;
  } else if (proposedIsUnchanged) {
    // no conflict - take changes from existing
    return existingFile;
  } else {
    // this case handles both false conflicts (where existing == proposed) and
    // true conflicts.
    if (options.binaryConflictPreference === 'existing') {
      return existingFile;
    } else {
      return proposedFile;
    }
  }
}

const BINARY_FILE_HEURISTIC_MAX_LENGTH = 8000;

/**
 * Returns whether the given Buffer is binary or not using a heuristic based
 * approach of checking for null bytes in the beginning of the file contents.
 */
export function isBinary(fileBuffer?: Buffer): boolean {
  if (!fileBuffer) {
    return false;
  }

  const searchLength = Math.min(fileBuffer.length, BINARY_FILE_HEURISTIC_MAX_LENGTH);
  for (let i = 0; i < searchLength; i++) {
    if (fileBuffer[i] === 0) {
      return true;
    }
  }

  return false;
}

function buffersEqual(a: Buffer | undefined, b: Buffer | undefined) {
  if (!a && !b) {
    return true;
  }

  if (!a || !b) {
    return false;
  }

  return a.equals(b);
}
