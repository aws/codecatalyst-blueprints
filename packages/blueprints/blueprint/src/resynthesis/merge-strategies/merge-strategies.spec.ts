import { readFileSync, readdirSync } from 'fs';
import * as path from 'path';
import { MergeStrategies } from './merge-strategies';
import { ContextFile } from '../context-file';

const EXAMPLE_ROOT_DIR = 'examples';
const EXAMPLE_DIRS = readdirSync(path.join(__dirname, EXAMPLE_ROOT_DIR));

[MergeStrategies.alwaysUpdate, MergeStrategies.neverUpdate, MergeStrategies.onlyAdd, MergeStrategies.threeWayMerge].forEach(mergeStrategy => {
  describe(`merge strategy: ${mergeStrategy.name}`, () => {
    it.each(EXAMPLE_DIRS)('matches snapshot for example: %s', example => {
      const { a, o, b } = getTestFiles(example);

      const commonAncestorFile: ContextFile = {
        buffer: o,
        path: 'file',
        repositoryName: 'repo',
      };

      const existingFile: ContextFile = {
        buffer: a,
        path: 'file',
        repositoryName: 'repo',
      };

      const proposedFile: ContextFile = {
        buffer: b,
        path: 'file',
        repositoryName: 'repo',
      };

      const resolvedFile = mergeStrategy(commonAncestorFile, existingFile, proposedFile);
      expect(resolvedFile?.buffer?.toString()).toMatchSnapshot();
    });
  });

  if (mergeStrategy === MergeStrategies.threeWayMerge) {
    it('returns the proposed file for binary files', () => {
      const ancestorBuffer = Buffer.from('Ancestor\x00Content');
      const existingBuffer = Buffer.from('Existing\x00Content');
      const proposedBuffer = Buffer.from('Proposed\x00Content');

      const [ancestorFile, existingFile, proposedFile] = [ancestorBuffer, existingBuffer, proposedBuffer].map(buffer => {
        return { buffer, path: 'file', repositoryName: 'repo' };
      });

      const resolvedFile = mergeStrategy(ancestorFile, existingFile, proposedFile);
      expect(resolvedFile?.buffer).toEqual(proposedBuffer);
    });
  }
});

function getTestFiles(dir: string): { a: Buffer; o: Buffer; b: Buffer } {
  const files = readdirSync(path.join(__dirname, EXAMPLE_ROOT_DIR, dir));

  const a = files.find(f => f.startsWith('a.'));
  const b = files.find(f => f.startsWith('b.'));
  const o = files.find(f => f.startsWith('o.'));

  if (!a || !b || !o) {
    throw new Error(`Failed to find test files in directory: ${dir}`);
  }

  return {
    a: readFileSync(path.join(__dirname, EXAMPLE_ROOT_DIR, dir, a)),
    b: readFileSync(path.join(__dirname, EXAMPLE_ROOT_DIR, dir, b)),
    o: readFileSync(path.join(__dirname, EXAMPLE_ROOT_DIR, dir, o)),
  };
}
