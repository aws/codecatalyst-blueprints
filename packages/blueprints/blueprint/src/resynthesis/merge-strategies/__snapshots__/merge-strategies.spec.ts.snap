// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge strategy: alwaysUpdate matches snapshot for example: empty-a 1`] = `
"hello world b
"
`;

exports[`merge strategy: alwaysUpdate matches snapshot for example: java 1`] = `
"// Sample file used in merge strategy tests:
package helloworld;

import java.util.HashMap;
import java.util.Map;

public class App {
    public static void main(String[] args) {
        for (int i = 0; i < args.length; i++) {
            Map<Character, Integer> result = countCharacters(args[i]);
            System.out.println(result.getOrDefault('b', 0));
        }

        System.out.println("Finished");
    }
    
    public static Map<Character, Integer> countCharacters(final String input) {
        Map<Character, Integer> result = new HashMap<>();
        for (int i = 0; i < input.length(); i++) {
            char c = Character.toLowerCase(input.charAt(i));
            result.put(c, result.getOrDefault(c, 0) + 1);
        }

        // insert a comment:
        return result;
    }
}
"
`;

exports[`merge strategy: alwaysUpdate matches snapshot for example: no-conflict 1`] = `
"first line
second line
"
`;

exports[`merge strategy: alwaysUpdate matches snapshot for example: no-diff 1`] = `
"this is the same line of text
"
`;

exports[`merge strategy: alwaysUpdate matches snapshot for example: simple 1`] = `
"modified in b
"
`;

exports[`merge strategy: neverUpdate matches snapshot for example: empty-a 1`] = `""`;

exports[`merge strategy: neverUpdate matches snapshot for example: java 1`] = `
"// Sample file used in merge strategy tests:
package helloworld;

import java.util.HashMap;
import java.util.Map;

public class App {
    public static void main(String[] args) {
        System.out.println("Starting...");

        for (int i = 0; i < args.length; i++) {
            Map<Character, Integer> result = countCharacters(args[i]);
            System.out.println(result.getOrDefault('a', 0));
        }

        System.out.println("Finished");
    }
    
    // Counts characters
    public static Map<Character, Integer> countCharacters(final String input) {
        Map<Character, Integer> result = new HashMap<>();
        for (int i = 0; i < input.length(); i++) {
            char c = Character.toUpperCase(input.charAt(i));
            result.put(c, result.getOrDefault(c, 0) + 1);
        }

        return result;
    }
}
"
`;

exports[`merge strategy: neverUpdate matches snapshot for example: no-conflict 1`] = `
"first line
inserted from a
second line
third line
"
`;

exports[`merge strategy: neverUpdate matches snapshot for example: no-diff 1`] = `
"this is the same line of text
"
`;

exports[`merge strategy: neverUpdate matches snapshot for example: simple 1`] = `
"modified in a
"
`;

exports[`merge strategy: onlyAdd matches snapshot for example: empty-a 1`] = `""`;

exports[`merge strategy: onlyAdd matches snapshot for example: java 1`] = `
"// Sample file used in merge strategy tests:
package helloworld;

import java.util.HashMap;
import java.util.Map;

public class App {
    public static void main(String[] args) {
        System.out.println("Starting...");

        for (int i = 0; i < args.length; i++) {
            Map<Character, Integer> result = countCharacters(args[i]);
            System.out.println(result.getOrDefault('a', 0));
        }

        System.out.println("Finished");
    }
    
    // Counts characters
    public static Map<Character, Integer> countCharacters(final String input) {
        Map<Character, Integer> result = new HashMap<>();
        for (int i = 0; i < input.length(); i++) {
            char c = Character.toUpperCase(input.charAt(i));
            result.put(c, result.getOrDefault(c, 0) + 1);
        }

        return result;
    }
}
"
`;

exports[`merge strategy: onlyAdd matches snapshot for example: no-conflict 1`] = `
"first line
inserted from a
second line
third line
"
`;

exports[`merge strategy: onlyAdd matches snapshot for example: no-diff 1`] = `
"this is the same line of text
"
`;

exports[`merge strategy: onlyAdd matches snapshot for example: simple 1`] = `
"modified in a
"
`;

exports[`merge strategy: threeWayMerge matches snapshot for example: empty-a 1`] = `
"<<<<<<< existing

=======
hello world b
>>>>>>> proposed
"
`;

exports[`merge strategy: threeWayMerge matches snapshot for example: java 1`] = `
"// Sample file used in merge strategy tests:
package helloworld;

import java.util.HashMap;
import java.util.Map;

public class App {
    public static void main(String[] args) {
        System.out.println("Starting...");

        for (int i = 0; i < args.length; i++) {
            Map<Character, Integer> result = countCharacters(args[i]);
<<<<<<< existing
            System.out.println(result.getOrDefault('a', 0));
=======
            System.out.println(result.getOrDefault('b', 0));
>>>>>>> proposed
        }

        System.out.println("Finished");
    }
    
    // Counts characters
    public static Map<Character, Integer> countCharacters(final String input) {
        Map<Character, Integer> result = new HashMap<>();
        for (int i = 0; i < input.length(); i++) {
<<<<<<< existing
            char c = Character.toUpperCase(input.charAt(i));
=======
            char c = Character.toLowerCase(input.charAt(i));
>>>>>>> proposed
            result.put(c, result.getOrDefault(c, 0) + 1);
        }

        // insert a comment:
        return result;
    }
}
"
`;

exports[`merge strategy: threeWayMerge matches snapshot for example: no-conflict 1`] = `
"first line
inserted from a
second line
"
`;

exports[`merge strategy: threeWayMerge matches snapshot for example: no-diff 1`] = `
"this is the same line of text
"
`;

exports[`merge strategy: threeWayMerge matches snapshot for example: simple 1`] = `
"<<<<<<< existing
modified in a
=======
modified in b
>>>>>>> proposed
"
`;
